builtin/credential/aws/path_role.go

```diff
@@ -486,37 +551,41 @@ func (b *backend) pathRoleCreateUpdate(ctx context.Context, req *logical.Request
 	}
 
 	if boundIamRoleARNRaw, ok := data.GetOk("bound_iam_role_arn"); ok {
-		roleEntry.BoundIamRoleARN = boundIamRoleARNRaw.(string)
+		roleEntry.BoundIamRoleARNs = boundIamRoleARNRaw.([]string)
 	}
 
 	if boundIamInstanceProfileARNRaw, ok := data.GetOk("bound_iam_instance_profile_arn"); ok {
-		roleEntry.BoundIamInstanceProfileARN = boundIamInstanceProfileARNRaw.(string)
+		roleEntry.BoundIamInstanceProfileARNs = boundIamInstanceProfileARNRaw.([]string)
 	}
 
 	if boundIamPrincipalARNRaw, ok := data.GetOk("bound_iam_principal_arn"); ok {
-		principalARN := boundIamPrincipalARNRaw.(string)
-		roleEntry.BoundIamPrincipalARN = principalARN
-		// Explicitly not checking to see if the user has changed the ARN under us
-		// This allows the user to sumbit an update with the same ARN to force Vault
-		// to re-resolve the ARN to the unique ID, in case an entity was deleted and
-		// recreated
-		if roleEntry.ResolveAWSUniqueIDs && roleEntry.BoundIamPrincipalARN != "" && !strings.HasSuffix(roleEntry.BoundIamPrincipalARN, "*") {
-			principalID, err := b.resolveArnToUniqueIDFunc(ctx, req.Storage, principalARN)
-			if err != nil {
-				return logical.ErrorResponse(fmt.Sprintf("failed updating the unique ID of ARN %#v: %#v", principalARN, err)), nil
+		principalARNs := boundIamPrincipalARNRaw.([]string)
+		roleEntry.BoundIamPrincipalARNs = principalARNs
+		roleEntry.BoundIamPrincipalIDs = []string{}
+		for _, principalARN := range principalARNs {
+			// Explicitly not checking to see if the user has changed the ARN under us
+			// This allows the user to sumbit an update with the same ARN to force Vault
+			// to re-resolve the ARN to the unique ID, in case an entity was deleted and
+			// recreated
+			if roleEntry.ResolveAWSUniqueIDs && !strings.HasSuffix(principalARN, "*") {
+				principalID, err := b.resolveArnToUniqueIDFunc(ctx, req.Storage, principalARN)
+				if err != nil {
+					return logical.ErrorResponse(fmt.Sprintf("failed updating the unique ID of ARN %#v: %#v", principalARN, err)), nil
+				}
+				roleEntry.BoundIamPrincipalIDs = append(roleEntry.BoundIamPrincipalIDs, principalID)
 			}
-			roleEntry.BoundIamPrincipalID = principalID
-		} else {
-			// Need to handle the case where we're switching from a non-wildcard principal to a wildcard principal
-			roleEntry.BoundIamPrincipalID = ""
```

Comment:
Is this case still being handled? Does it need to be?

Reply:
I believe it does the right thing.

Before, the code needed to handle switching from non-widlcard to a wildcard. Now, however, it needs to handle switching from a list containing some widlcard and some non-wildcard ARNs, to a different list containing some (possibly same and possibly different) wildcard ARNs and some (possibly same and possibly different) non-wildcard ARNs. So this is no longer a special case.

It's handled by calling `roleEntry.BoundIamPrincipalIDs = []string{}` whenever `bound_iam_principal_arn` is explicitly set and then building up the `BoundIamPrincipalIDs` for each non-wildcard principal.